### ISSUE TITLE
App running in background, reliably and debuggable

### ISSUE THREAD

**Issue by scripting** [scripting/Scripting-News#138]
TL;DR: I am working on a Node app that must run in the background, disconnected from the terminal, survives the terminal window being closed, yet is easy to quit so I can run a new version. It's a product that will remain in active development at the same time it is deployed. 

I am currently using `nohup node appname &` to launch the app, and `killAll node`, to terminate it, and the apps it launches so I can relaunch. 

This method seems pretty good. I am able to Control-C after launching the app and it keeps running in the background. It seems to survive closing the terminal and logging on again.

Curious to know what the braintrust thinks? What method do you use to launch and kill an app in the background when you're actively developing it?

**Comment by easp** [comment]
Sounds like you are running on a unix variant. Which one? They have different canonical ways to manage background daemons/servers.

**Comment by scripting** [comment]
Ooops. I should have said -- Ubuntu.

![image](https://user-images.githubusercontent.com/1686843/68884459-3d0ea680-06e1-11ea-90cd-ccb02af67271.png)

**Comment by toontoet** [comment]
Have a look at what pm2 has to offer. It allows you to manage node processes and more: https://pm2.keymetrics.io/

It also has a nice develop feature so it restarts an app when source code changes. And nice features to deploy code to remote servers.

**Comment by scripting** [comment]
Heh. That's kind of what I'm working on. ;-)

**Comment by mterenzio** [comment]
Will `forever` keep it running in the background?

Then `forever list`

And `forever stop [pid]`

**Comment by scripting** [comment]
Matt. this app *is* forever.

**Comment by tedchoward** [comment]
I’ve been using [pm2](https://www.npmjs.com/package/pm2) on my server to run river5 for quite a while. It keeps it going, survives restarts, etc. I haven’t thought a lot about it, I just set it up once a long time ago, and then forgot about it.

**Comment by danmactough** [comment]
> I am currently using `nohup node appname.js &` to launch the app, and `killAll node`, to terminate it, and the apps it launches.

With your requirements, this is what I would do. (Actually, I run everything inside Docker containers, now. But before I started doing that, this is what I would have done.)

**Comment by easp** [comment]
I'd use "systemd" to manage the node process. That way it gets started when/if the machine reboots. It can also automatically restart the process after a crash. Systemd is already installed on ubuntu and is responsible for managing pretty much every persistent process and subsystem, so you won't have to worry about installing extra software.

[This article](https://nodesource.com/blog/running-your-node-js-app-with-systemd-part-1/) looks like a decent overview of configuring a systemd managed node app.

**Comment by scripting** [comment]
Thanks @easp -- I'll look into it. I see it does allow for the possibility of quitting and relaunching, so it may be usable for a project under development like this one.
